cmd/tailscale/cli/web.go

```diff
@@ -247,7 +247,13 @@ func qnapAuthnSid(r *http.Request, user, sid string) (string, *qnapAuthResponse,
 }
 
 func qnapAuthnFinish(user, url string) (string, *qnapAuthResponse, error) {
-	resp, err := http.Get(url)
+	// QNAP Force HTTPS mode uses a self-signed certificate.
+	// https://github.com/tailscale/tailscale/issues/6903
```

Comment:
Can we somehow discover the cert and use that here?

Reply:
It currently lives in /etc/ssl/certs/myhost.crt (and another copy of itself in /etc/ssl/certs/myrootca.crt to sign itself with).

I was concerned about the fragility in case QNAP ever moves it, but I suppose it isn't likely to move outside of /etc/ssl.

Reply:
I tried to do this, by importing the QNAP root certificate:
```
+func qnapGetInsecureTLSConfig() *tls.Config {
+       return &tls.Config{InsecureSkipVerify: true}
+}
+
+func qnapGetTLSConfig() *tls.Config {
+       rootCAs, err := x509.SystemCertPool()
+       if err != nil {
+               log.Printf("qnapGetTLSConfig SystemCertPool failed: %v, falling back to no verification", err)
+               return qnapGetInsecureTLSConfig()
+       }
+
+       pem, err := os.ReadFile("/etc/ssl/certs/myrootca.crt")
+       if err != nil {
+               log.Printf("qnapGetTLSConfig read myrootca.crt: %v, falling back to no verification", err)
+               return qnapGetInsecureTLSConfig()
+       }
+
+       if !rootCAs.AppendCertsFromPEM(pem) {
+               log.Println("qnapGetTLSConfig AppendCertsFromPEM failed, falling back to no verification", err)
+               return qnapGetInsecureTLSConfig()
+       }
+
+       return &tls.Config{RootCAs: rootCAs}
+}
+
 func qnapAuthnFinish(user, url string) (string, *qnapAuthResponse, error) {
        // QNAP Force HTTPS mode uses a self-signed certificate.
        // https://github.com/tailscale/tailscale/issues/6903
        tr := &http.Transport{
-               TLSClientConfig: &tls.Config{InsecureSkipVerify: true},
+               TLSClientConfig: qnapGetTLSConfig(),
        }
        client := &http.Client{Transport: tr}
        resp, err := client.Get(url)
```

However this still doesn't work because though it is now signed by a valid root certificate, none of the CNs match and there are no alternative names:

`Get "https://127.0.0.1:443/cgi-bin/authLogin.cgi?sid=jvys7c89": x509: cannot validate certificate for 127.0.0.1 because it doesn't contain any IP SANs`

The fields in the certificate are:
```
emailAddress = support@qnap.com
CN = QNAP NAS
OU = QTS
O = QNAP Systems, Inc.
L = Taipei
ST = Taipei
C = TW
```

The CN is not a valid hostname, and there are no alternate names.
I don't think this certificate can pass any validation.

Reply:
Thanks for testing. Can you leave a comment describing this in the code?